routing/router.go

```diff
@@ -2163,15 +2164,29 @@ func (r *ChannelRouter) SendToRoute(htlcHash lntypes.Hash, rt *route.Route) (
 	// mark the payment failed with the control tower immediately. Process
 	// the error to check if it maps into a terminal error code, if not use
 	// a generic NO_ROUTE error.
-	if err := sh.handleSendError(attempt, shardError); err != nil {
-		return nil, err
-	}
+	var failureReason *channeldb.FailureReason
+	err = sh.handleSendError(attempt, shardError)
 
-	err = r.cfg.Control.Fail(
-		paymentIdentifier, channeldb.FailureReasonNoRoute,
-	)
-	if err != nil {
-		return nil, err
+	switch {
+	// If we weren't able to extract a proper failure reason (which can
+	// happen if the second chance logic is triggered), then we'll use the
+	// normal no route error.
+	case err == nil:
+		err = r.cfg.Control.Fail(
+			paymentIdentifier, channeldb.FailureReasonNoRoute,
+		)
+		if err != nil {
```

Comment:
The error check could be moved out of the switch statements, saving us a few lines.